msm/mycroft_skills_manager.py

```diff
@@ -46,20 +56,139 @@ def __init__(self, platform='default', skills_dir=None, repo=None,
+                entry['update'] = \
+                    skills_data.get(skill.name, {}).get('updated') or 0
+
+                new['skills'].append(entry)
+            new['upgraded'] = True
+        return new
+
+    def curate_skills_data(self, skills_data):
+        """ Sync skills_data with actual skills on disk. """
+        local_skills = [s for s in self.list() if s.is_local]
+        default_skills = [s.name for s in self.list_defaults()]
+        local_skill_names = [s.name for s in local_skills]
+        skills_data_skills = [s['name'] for s in skills_data['skills']]
+
+        # Check for skills that aren't in the list
+        for skill in local_skills:
+            if skill.name not in skills_data_skills:
+                if skill.name in default_skills:
+                    origin = 'default'
+                elif skill.url:
+                    origin = 'cli'
+                else:
+                    origin = 'non-msm'
+                entry = build_skill_entry(skill.name, origin, False)
+                skills_data['skills'].append(entry)
 
-    def install(self, param, author=None, constraints=None):
+        # Check for skills in the list that doesn't exist in the filesystem
+        remove_list = []
+        for s in skills_data.get('skills', []):
+            if (s['name'] not in local_skill_names and
+                    s['installation'] == 'installed'):
+                remove_list.append(s)
+        for skill in remove_list:
+            skills_data['skills'].remove(skill)
+        return skills_data
+
+    def load_skills_data(self) -> dict:
+        skills_data = load_skills_data()
+        if skills_data.get('version', 0) < CURRENT_SKILLS_DATA_VERSION:
+            skills_data = self.__upgrade_skills_data(skills_data)
+        else:
+            skills_data = self.curate_skills_data(skills_data)
+        return skills_data
+
+    def sync_skills_data(self):
+        """ Update internal skill_data_structure from disk. """
+        self.skills_data = self.load_skills_data()
+        if 'upgraded' in self.skills_data:
+            self.skills_data.pop('upgraded')
+            self.skills_data_hash = ''
+        else:
+            self.skills_data_hash = skills_data_hash(self.skills_data)
+
+    def write_skills_data(self, data=None):
+        """ Write skills data hash if it has been modified. """
+        data = data or self.skills_data
+        if skills_data_hash(data) != self.skills_data_hash:
+            write_skills_data(data)
+
+    def install(self, param, author=None, constraints=None, origin=''):
         """Install by url or name"""
-        self.find_skill(param, author).install(constraints)
+        if isinstance(param, SkillEntry):
+            skill = param
+        else:
+            skill = self.find_skill(param, author)
+        entry = build_skill_entry(skill.name, origin, skill.is_beta)
+        try:
+            skill.install(constraints)
+            entry['installed'] = time.time()
```

Comment:
Just an idea, could be cleaner via:
```python
entry.update(installed=time.time(), installation='installed', status='activate')
```
What do you think?

Reply:
That does look nice.